react/src/hooks/utils.tsx

```diff
@@ -13,24 +13,54 @@ import {
 } from "react-query";
 import { SetDataOptions } from "react-query/types/core/query";
 
-interface FetchOptions {
-    onSuccess?: (res: Response) => any;
-    onError?: (res: Response) => any;
-}
-
 /**
  * Fetch the provided url. Return the JSON response if successful.
  * Throw the response if unsuccessful.
  */
 export async function basicFetch(
     url: string,
-    params: Record<string, string> = {},
-    options?: FetchOptions
+    params: Record<string, any> = {},
+    options: RequestInit | undefined = undefined
 ) {
     const paramString = Object.keys(params).length ? `?${new URLSearchParams(params)}` : "";
-    const response = await fetch(`${url}${paramString}`);
+    const response = await fetch(`${url}${paramString}`, options);
     if (response.ok) {
-        return response.json();
+        if (response.headers.get("Content-Type") === "text/csv") {
+            return response.text();
+        } else return response.json();
+    } else {
+        throw response;
+    }
+}
+
+/**
+ * Fetch csv from provided url. Download csv if successful.
+ * Throw the response if unsuccessful.
+ */
+export async function fetchAndDownloadCsv(
+    url: string,
+    params: Record<string, any> = {},
+    options: RequestInit | undefined = {}
+) {
+    let headers = { Accept: "text/csv" };
+    if (options.headers) {
+        headers = { ...options.headers, ...headers };
+    }
+    const paramString = Object.keys(params).length ? `?${new URLSearchParams(params)}` : "";
+    const response = await fetch(`${url}${paramString}`, { ...options, headers });
+    if (response.ok) {
+        const csv = await response.text();
+        const downloadLink = document.createElement("a");
+        const filename = (response.headers.get("content-disposition") || "foo=report.csv").replace(
+            /.+=/,
+            ""
+        );
+        const url = URL.createObjectURL(new Blob(["\ufeff", csv]));
+        downloadLink.href = url;
+        downloadLink.download = filename;
+        document.body.appendChild(downloadLink);
+        downloadLink.click();
+        document.body.removeChild(downloadLink);
```

Comment:
I was hoping that this approach wouldn't be needed if we already have a content-disposition header on the backend. `window.open` might get blocked by a pop-up blocker, but would directly navigating to the endpoint work?

Reply:
I don't think so -- afaik if we're making an async fetch the browser won't pay attention to the content-disposition header. We could navigate the user to the url if the url held a static file (e.g., the endpoint doesn't return the csv but saves it somewhere publicly accessible and returns the path and then we point the browser there), but if the file comes from an api route, redirecting the user there will just reload the app and the router will take over.

Reply:
Is it because this is used in a react-query hook rather than just being a standalone side effect function? Or is the problem that we're distinguishing the return type based on the Accept header rather than something in the URL, which would allow it to be used as a link?

Reply:
The latter, I think. Explanation can be found [here](https://create-react-app.dev/docs/proxying-api-requests-in-development/) in terms of current setup. If we point the browser at, say, `stager.local/api/download-some-file`, the server will see that the browser request's Accept header lists text/html before */* and respond with index.html, and then react will navigate to `/api/download-some-file`, which will display whatever view the router defaults to rendering at that path.

Reply:
I see, the problem is that it wouldn't work against the development server, though it would work in production. This never came up as we don't open the endpoints in the browser, which would send `Accept: text/html; other stuff; */*`. `fetch` and `curl` both use `*/*`.